File: week1/Design_pattern&principles/exercise1/SingletonPatternExample/Logger.java
public class Logger {
    // Step 1: Private static instance of the class
    private static Logger instance;

    // Step 2: Private constructor to prevent instantiation
    private Logger() {
        System.out.println("Logger initialized");
    }

    // Step 3: Public static method to return the single instance
    public static Logger getInstance() {
        if (instance == null) {
            instance = new Logger();
        }
        return instance;
    }

    // Example log method
    public void log(String message) {
        System.out.println("LOG: " + message);
    }
}

File: week1/Design_pattern_principles/exercise1/Logger.cs
using System;

public sealed class Logger
{
    private static readonly object lockObj = new object();
    private static Logger? instance = null;

    private Logger()
    {
        Console.WriteLine("Logger initialized");
    }

    public static Logger GetInstance()
    {
        if (instance == null)
        {
            lock (lockObj)
            {
                if (instance == null)
                {
                    instance = new Logger();
                }
            }
        }
        return instance;
    }

    public void Log(string message)
    {
        Console.WriteLine("LOG: " + message);
    }
}
